doc/sphinx/addendum/implicit-coercions.rst

```diff
@@ -82,6 +82,19 @@ To coerce an object :g:`t:C t₁..tₙ` of ``C`` towards ``D``, we have to
 apply the coercion ``f`` to it; the obtained term :g:`f _.._ t` is
 then an object of ``D``.
 
+Reverse Coercions
+-----------------
+
+When no coercion can be applied, reverse coercions are looked for.
+That is, when :g:`a : A'` is provided where something of type :g:`A`
+is expected, if there is a coercion :g:`f : A >-> A'` and some :g:`x :
+A` can be infered (typically through :ref:`canonicalstructures` or
+:ref:`typeclasses`) such that :g:`f x = a`, then :g:`a` is replaced by
+:g:`x`. See the :ref:`example below <example-reverse-coercion>`. If no
+:g:`f : A >-> A'` can be found, a coercion :g:`g : A' >-> A''` is
+looked for such that there are a coercion :g:`f : A >-> A''` and some
+:g:`x : A` for which :g:`f x = g a`. Only the coercions declared with
+the :attr:`reverse` attribute can be used as functions :g:`f` above.
```

Comment:
I'm pretty sure Jim will do abetter job, but in the meanwhile here my 2 C:
```suggestion
A term is reverse-coerced only if it cannot be coerced (directly) to its expect type.
A coercion is said to be reversible if it is declared with with the :attr:`reverse` attribute.

In order to cast a term :g:`a : A` to type :g:`B` using a reverse coercion we:

1. find a reversible coercion :g:`f : B >-> A`

2. synthesize  some :g:`?x : B` such that :g:`f ?x = a`,
    typically through :ref:`canonicalstructures` or :ref:`typeclasses`
    
3. replace :g:`a` with the value of :g:`?x`

If there is no such reversible coercion :g:`f : B >-> A`, then we:

1. find a midpoint :g:`C` equipped with an incoming reversible coercion :g:`g : B >-> C` and
    a coercion :g:`h : A >-> C` (not necessarily reversible)
 
2. synthesize  some :g:`?x : B` such that :g:`g ?x = h a`

3. we replace :g:`a` with the value of :g:`?x`

If more than one midpoint :g:`C` exists, we try the procedure with all of them in the following order:
a midpoint :g:`C` is tried before a midpoint :g:`D` if these exists a coercion from :g:`C` to :g:`D`.
The intuition is that since coercions forget information, :g:`D` has less information that :g:`C`, and hence
inferring :g:`?x : B` from less information would be harder.

See the :ref:`example below <example-reverse-coercion>`.
```

Reply:
@proux01 what do you think of this suggestion? Once we mark the PR a ready for review doc-maintainers will be notified, so this suggestion should be accepted/rejected before that.

Reply:
LGTM, thanks!